File: android/worker/app/src/main/java/com/workforcex/worker/ui/worker/BrowseJobsActivity.java
package com.workforcex.worker.ui.worker;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.workforcex.worker.api.JobApplication;
import com.workforcex.worker.api.JobBrowseItem;
import com.workforcex.worker.api.RetrofitClient;
import com.workforcex.worker.databinding.ActivityBrowseJobsBinding;
import com.workforcex.worker.utils.TokenManager;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class BrowseJobsActivity extends AppCompatActivity {

    private ActivityBrowseJobsBinding binding;
    private TokenManager tokenManager;
    private BrowseJobsAdapter adapter;
    private List<JobBrowseItem> allJobs = new ArrayList<>();
    private String filterSkills;
    private String filterCity;
    private boolean showMatchedFirst;
    private boolean currentlyShowingMatched = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityBrowseJobsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        tokenManager = new TokenManager(this);
        adapter = new BrowseJobsAdapter(new ArrayList<>(), this::applyToJob);
        binding.rvJobs.setLayoutManager(new LinearLayoutManager(this));
        binding.rvJobs.setAdapter(adapter);

        filterSkills     = getIntent().getStringExtra("filterSkills");
        filterCity       = getIntent().getStringExtra("filterCity");
        showMatchedFirst = getIntent().getBooleanExtra("showMatchedFirst", false);

        if (showMatchedFirst && filterSkills != null && !filterSkills.isEmpty()) {
            setTitle("Jobs Matching Your Skills");
            binding.btnToggleJobs.setVisibility(View.VISIBLE);
            binding.btnToggleJobs.setText("Show All Jobs");
            binding.tvFilterInfo.setVisibility(View.VISIBLE);
            binding.tvFilterInfo.setText("Matching: " + filterSkills);
        } else {
            setTitle("Available Jobs");
            binding.btnToggleJobs.setVisibility(View.GONE);
            binding.tvFilterInfo.setVisibility(View.GONE);
            currentlyShowingMatched = false;
        }

        binding.btnToggleJobs.setOnClickListener(v -> toggleJobsView());
        loadJobs();
    }

    private void loadJobs() {
        binding.progressBar.setVisibility(View.VISIBLE);
        RetrofitClient.get().browseJobs(tokenManager.getBearerToken())
                .enqueue(new Callback<List<JobBrowseItem>>() {
                    @Override
                    public void onResponse(Call<List<JobBrowseItem>> call,
                                           Response<List<JobBrowseItem>> response) {
                        binding.progressBar.setVisibility(View.GONE);
                        if (response.isSuccessful() && response.body() != null) {
                            allJobs = response.body();
                            displayJobs(currentlyShowingMatched && showMatchedFirst);
                        }
                    }
                    @Override
                    public void onFailure(Call<List<JobBrowseItem>> call, Throwable t) {
                        binding.progressBar.setVisibility(View.GONE);
                        Toast.makeText(BrowseJobsActivity.this,
                                "Network error: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void toggleJobsView() {
        currentlyShowingMatched = !currentlyShowingMatched;
        displayJobs(currentlyShowingMatched);
        binding.btnToggleJobs.setText(currentlyShowingMatched ? "Show All Jobs" : "Show Matching Jobs");
        binding.tvFilterInfo.setVisibility(currentlyShowingMatched ? View.VISIBLE : View.GONE);
        setTitle(currentlyShowingMatched ? "Jobs Matching Your Skills" :
                "All Jobs (" + allJobs.size() + ")");
    }

    private void displayJobs(boolean matchedOnly) {
        List<JobBrowseItem> toShow = matchedOnly && filterSkills != null && !filterSkills.isEmpty()
                ? filterBySkills(allJobs, filterSkills, filterCity)
                : allJobs;
        adapter.update(toShow);
        binding.tvEmpty.setVisibility(toShow.isEmpty() ? View.VISIBLE : View.GONE);
        binding.tvResultCount.setVisibility(View.VISIBLE);
        binding.tvResultCount.setText(toShow.size() +
                (matchedOnly ? " job(s) match your skills" : " total jobs available"));
    }

    private void applyToJob(JobBrowseItem job, int position) {
        RetrofitClient.get().applyToJob(tokenManager.getBearerToken(), job.id)
                .enqueue(new Callback<JobApplication>() {
                    @Override
                    public void onResponse(Call<JobApplication> call,
                                           Response<JobApplication> response) {
                        if (response.isSuccessful()) {
                            Toast.makeText(BrowseJobsActivity.this,
                                    "Applied to " + job.title + "! ✓", Toast.LENGTH_SHORT).show();
                        } else if (response.code() == 400) {
                            Toast.makeText(BrowseJobsActivity.this,
                                    "Already applied to this job", Toast.LENGTH_SHORT).show();
                        } else {
                            Toast.makeText(BrowseJobsActivity.this,
                                    "Failed to apply", Toast.LENGTH_SHORT).show();
                        }
                    }
                    @Override
                    public void onFailure(Call<JobApplication> call, Throwable t) {
                        Toast.makeText(BrowseJobsActivity.this,
                                "Network error", Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private List<JobBrowseItem> filterBySkills(List<JobBrowseItem> jobs, String skills, String city) {
        Set<String> workerSkills = new HashSet<>(Arrays.asList(skills.toLowerCase().split(",")));
        return jobs.stream().filter(job -> {
            if (job.skillsRequired == null) return false;
            Set<String> jobSkills = new HashSet<>(Arrays.asList(job.skillsRequired.toLowerCase().split(",")));
            boolean match = jobSkills.stream().anyMatch(workerSkills::contains);
            if (!match) return false;
            if (city != null && !city.isEmpty() && job.location != null)
                return job.location.toLowerCase().contains(city.toLowerCase());
            return true;
        }).collect(Collectors.toList());
    }
}

File: android/worker/app/src/main/java/com/workforcex/worker/ui/worker/BrowseJobsActivity.kt
package com.workforcex.worker.ui.worker

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.workforcex.shared.models.JobApplication
import com.workforcex.shared.models.JobBrowseItem
import com.workforcex.shared.RetrofitClient
import com.workforcex.worker.databinding.ActivityBrowseJobsBinding
import com.workforcex.worker.utils.TokenManager
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*
import java.util.stream.Collectors

class BrowseJobsActivity : AppCompatActivity() {

    private lateinit var binding: ActivityBrowseJobsBinding
    private lateinit var tokenManager: TokenManager
    private lateinit var adapter: BrowseJobsAdapter
    private var allJobs: List<JobBrowseItem> = ArrayList()
    private val appliedJobIds: MutableSet<String> = HashSet()
    private var filterSkills: String? = null
    private var filterCity: String? = null
    private var showMatchedFirst = false
    private var currentlyShowingMatched = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBrowseJobsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        tokenManager = TokenManager(this)
        adapter = BrowseJobsAdapter(ArrayList(), this::applyToJob, appliedJobIds)
        binding.rvJobs.layoutManager = LinearLayoutManager(this)
        binding.rvJobs.adapter = adapter

        filterSkills = intent.getStringExtra("filterSkills")
        filterCity = intent.getStringExtra("filterCity")
        showMatchedFirst = intent.getBooleanExtra("showMatchedFirst", false)

        if (showMatchedFirst && !filterSkills.isNullOrEmpty()) {
            title = "Jobs Matching Your Skills"
            binding.btnToggleJobs.visibility = View.VISIBLE
            binding.btnToggleJobs.text = "Show All Jobs"
            binding.tvFilterInfo.visibility = View.VISIBLE
            binding.tvFilterInfo.text = "Matching: $filterSkills"
        } else {
            title = "Available Jobs"
            binding.btnToggleJobs.visibility = View.GONE
            binding.tvFilterInfo.visibility = View.GONE
            currentlyShowingMatched = false
        }

        binding.btnToggleJobs.setOnClickListener { toggleJobsView() }
        loadInitialData()
    }

    private fun loadInitialData() {
        binding.progressBar.visibility = View.VISIBLE
        // First, get the jobs this worker has already applied to
        RetrofitClient.get().getMyApplications(tokenManager.getBearerToken())
            .enqueue(object : Callback<List<JobApplication>> {
                override fun onResponse(
                    call: Call<List<JobApplication>>,
                    response: Response<List<JobApplication>>
                ) {
                    if (response.isSuccessful && response.body() != null) {
                        appliedJobIds.clear()
                        val newIds = response.body()!!.stream()
                            .map { app: JobApplication -> app.jobId }
                            .collect(Collectors.toSet())
                        appliedJobIds.addAll(newIds)
                    }
                    // Then, load all jobs
                    loadJobs()
                }

                override fun onFailure(call: Call<List<JobApplication>>, t: Throwable) {
                    // Still load jobs even if this fails
                    loadJobs()
                }
            })
    }

    private fun loadJobs() {
        RetrofitClient.get().browseJobs(tokenManager.getBearerToken())
            .enqueue(object : Callback<List<JobBrowseItem>> {
                override fun onResponse(
                    call: Call<List<JobBrowseItem>>,
                    response: Response<List<JobBrowseItem>>
                ) {
                    binding.progressBar.visibility = View.GONE
                    if (response.isSuccessful && response.body() != null) {
                        allJobs = response.body()!!
                        displayJobs(currentlyShowingMatched && showMatchedFirst)
                    }
                }

                override fun onFailure(call: Call<List<JobBrowseItem>>, t: Throwable) {
                    binding.progressBar.visibility = View.GONE
                    Toast.makeText(
                        this@BrowseJobsActivity,
                        "Network error: " + t.message, Toast.LENGTH_SHORT
                    ).show()
                }
            })
    }

    private fun toggleJobsView() {
        currentlyShowingMatched = !currentlyShowingMatched
        displayJobs(currentlyShowingMatched)
        binding.btnToggleJobs.text = if (currentlyShowingMatched) "Show All Jobs" else "Show Matching Jobs"
        binding.tvFilterInfo.visibility = if (currentlyShowingMatched) View.VISIBLE else View.GONE
        title = if (currentlyShowingMatched) "Jobs Matching Your Skills" else "All Jobs (${allJobs.size})"
    }

    private fun displayJobs(matchedOnly: Boolean) {
        val toShow = if (matchedOnly && !filterSkills.isNullOrEmpty())
            filterBySkills(allJobs, filterSkills, filterCity)
        else
            allJobs
        adapter.update(toShow)
        binding.tvEmpty.visibility = if (toShow.isEmpty()) View.VISIBLE else View.GONE
        binding.tvResultCount.visibility = View.VISIBLE
        binding.tvResultCount.text = (toShow.size.toString() +
                if (matchedOnly) " job(s) match your skills" else " total jobs available")
    }

    private fun applyToJob(job: JobBrowseItem, position: Int) {
        // Disable button immediately to prevent double-clicks
        job.applied = true
        adapter.notifyItemChanged(position)

        RetrofitClient.get().applyToJob(tokenManager.getBearerToken(), job.id)
            .enqueue(object : Callback<JobApplication> {
                override fun onResponse(
                    call: Call<JobApplication>,
                    response: Response<JobApplication>
                ) {
                    if (response.isSuccessful) {
                        Toast.makeText(
                            this@BrowseJobsActivity,
                            "Applied to " + job.title + "! ✓", Toast.LENGTH_SHORT
                        ).show()
                        appliedJobIds.add(job.id) // Add to our set
                    } else {
                        Toast.makeText(
                            this@BrowseJobsActivity,
                            "Failed to apply", Toast.LENGTH_SHORT
                        ).show()
                        // Re-enable button if the call failed
                        job.applied = false
                        adapter.notifyItemChanged(position)
                    }
                }

                override fun onFailure(call: Call<JobApplication>, t: Throwable) {
                    Toast.makeText(
                        this@BrowseJobsActivity,
                        "Network error", Toast.LENGTH_SHORT
                    ).show()
                    // Re-enable button on failure
                    job.applied = false
                    adapter.notifyItemChanged(position)
                }
            })
    }

    private fun filterBySkills(jobs: List<JobBrowseItem>, skills: String?, city: String?): List<JobBrowseItem> {
        val workerSkills = HashSet(Arrays.asList(*skills!!.lowercase().split(",").toTypedArray()))
        return jobs.stream().filter { job: JobBrowseItem ->
            if (job.skillsRequired == null) return@filter false
            val jobSkills = HashSet(Arrays.asList(*job.skillsRequired.lowercase().split(",").toTypedArray()))
            val match = jobSkills.stream().anyMatch { s: String? -> workerSkills.contains(s) }
            if (!match) return@filter false
            if (!city.isNullOrEmpty() && job.location != null)
                return@filter job.location.lowercase(Locale.getDefault()).contains(city.lowercase(Locale.getDefault()))
            true
        }.collect(Collectors.toList())
    }
}